llvm/lib/Transforms/Scalar/SROA.cpp

```diff
@@ -3531,8 +3531,19 @@ class AllocaSliceRewriter : public InstVisitor<AllocaSliceRewriter, bool> {
       return true;
     }
 
-    if (II.isLaunderOrStripInvariantGroup())
+    if (II.isLaunderOrStripInvariantGroup()) {
+      Value *AdjustedPtr = getNewAllocaSlicePtr(IRB, OldPtr->getType());
+      Value *New = nullptr;
+      if (II.getIntrinsicID() == Intrinsic::launder_invariant_group)
+        New = IRB.CreateLaunderInvariantGroup(AdjustedPtr);
+      else if (II.getIntrinsicID() == Intrinsic::strip_invariant_group)
+        New = IRB.CreateStripInvariantGroup(AdjustedPtr);
+
+      New->takeName(&II);
+      II.replaceAllUsesWith(New);
```

Comment:
Is this RAUW correct? Can't the alloca be split into two parts, both of which need their own intrinsic?

Reply:
Isn't this already being done as part of visiting each slice in:

https://github.com/llvm/llvm-project/blob/e03f427196ec67a8a5cfbdd658f9eabe9bce83ce/llvm/lib/Transforms/Scalar/SROA.cpp#L4921-L4924

Or am I mistaking it? It looks like the intrinsic has been rewritten for each part of the alloca (this is before mem2reg for the test case):
```
define void @partial_promotion_of_alloca() {
  %struct_ptr.sroa.0 = alloca i32, align 4
  %struct_ptr.sroa.2 = alloca i32, align 4
  store i32 0, ptr %struct_ptr.sroa.0, align 4
  store volatile i32 0, ptr %struct_ptr.sroa.2, align 4
  %barr = call ptr @llvm.launder.invariant.group.p0(ptr %struct_ptr.sroa.0)
  %1 = call ptr @llvm.launder.invariant.group.p0(ptr %struct_ptr.sroa.2)
  %struct_ptr.sroa.2.0.struct_ptr.sroa.2.4.load_val = load volatile i32, ptr %struct_ptr.sroa.2, align 4
  ret void
}
```

Reply:
The case I have in mind is something like memcpy_after_laundering_alloca but where we end up with two memcpys rather than one. I expect you'll get incorrect pointers in that case.

Reply:
Would you mind showing an example that could lead to get incorrect pointers? I had in mind something as follows, where we memcpy each field separately, but I suspect this is not what you were referring to:
```llvm
define void @memcpy_after_laundering_allocas_field(ptr %ptr) {
  %alloca = alloca { i64, i64 }, align 8
  %field1_ptr = getelementptr inbounds { i64, i64 }, ptr %alloca, i32 0, i32 0
  %field2_ptr = getelementptr inbounds { i64, i64 }, ptr %alloca, i32 0, i32 1
  %launder_field1 = call ptr @llvm.launder.invariant.group.p0(ptr %field1_ptr)
  %launder_field2 = call ptr @llvm.launder.invariant.group.p0(ptr %field2_ptr)
  call void @llvm.memcpy.p0.p0.i64(ptr %launder_field1, ptr %ptr, i64 8, i1 false)
  call void @llvm.memcpy.p0.p0.i64(ptr %launder_field2, ptr %ptr, i64 8, i1 false)
  ret void
}
```

Reply:
```llvm
define void @test(ptr %ptr) {
  %alloca = alloca { [16 x i8], i64, [16 x i8] }, align 8
  %launder = call ptr @llvm.launder.invariant.group.p0(ptr %alloca)
  %gep = getelementptr i8, ptr %launder, i64 16
  store i64 0, ptr %gep
  call void @llvm.memcpy.p0.p0.i64(ptr %launder, ptr %ptr, i64 40, i1 false)
  ret void
}
```

With your patch this produces:
```
define void @test(ptr %ptr) {
  %alloca.sroa.0 = alloca [16 x i8], align 8
  %alloca.sroa.3 = alloca [16 x i8], align 8
  %1 = call ptr @llvm.launder.invariant.group.p0(ptr %alloca.sroa.3)
  call void @llvm.memcpy.p0.p0.i64(ptr align 8 %alloca.sroa.0, ptr align 1 %ptr, i64 16, i1 false)
  %alloca.sroa.2.0.ptr.sroa_idx = getelementptr inbounds i8, ptr %ptr, i64 16
  %alloca.sroa.2.0.copyload = load i64, ptr %alloca.sroa.2.0.ptr.sroa_idx, align 1
  %alloca.sroa.3.0.ptr.sroa_idx = getelementptr inbounds i8, ptr %ptr, i64 24
  call void @llvm.memcpy.p0.p0.i64(ptr align 8 %alloca.sroa.3, ptr align 1 %alloca.sroa.3.0.ptr.sroa_idx, i64 16, i1 false)
  ret void
}
```
Note that the `launder.invariant` intrinsic here is unused.

Reply:
Updated this by considering whether we need to rewrite memory intrinsics as well when rewriting the invariant group one. Tried to keep everything self-contained in visitIntrinsicInst, but code was getting duplicated fast, and the memcpys were later being visited anyways. Thus, as the alloca may get split at each AllocaSliceRewriter iteration, if some of the users of the invariant group are memory intrinsics, we track the launder so as to use this one as ptr destination, instead of the new alloca.